### PR TITLE
Introduce tiny tuples

### PR DESCRIPTION
**Pull request by PersDep**

### CODE REVIEW

**Review by alyapunov — CHANGES_REQUESTED** [review]
1. The main part of the patch.
1.1. I don't like but I can live with `bool is_type` added to lots of functions. But the necessity of tracking the size of tuple manually seems to be unacceptable. See `vy_stmt_set_flags` method, it became a crap. Also I'm afraid of UB.
1.2. A least getters and setters of `bsize` should be extracted to separate commit.
2. The test
2.1. There's huge amount of bugs.
2.2. The proposed test subsystem (perf.h) is wrong and unusable.
2.3. The test itself is chaotic and hard to understand.
2.4. It does not test:
2.4.1. Performance of accessing tuple members (like bsize) in different workloads - tiny, big and mix.
2.4.2. Performance of access to indexed tuple fields. Current test of field access is inappropriate, it tests the speed of  `mp_next`.

File: cmake/compiler.cmake [hunk 1]
```diff
@@ -348,7 +348,7 @@ macro(enable_tnt_compile_flags)
         add_compile_flags("CXX"
             "-Wno-invalid-offsetof"
         )
-        add_compile_flags("C;CXX" "-Wno-gnu-alignof-expression")
+        add_compile_flags("C;CXX" "-Wno-gnu-alignof-expression -Wno-gnu-variable-sized-type-not-at-end")
```
[the remark this answers is not in the document]
Comment (PersDep) [reply cmake/compiler.cmake:351]: The warning is introduced by the `struct tuple` of variable size at the start of struct vy_stmt. It is now solved, struct vy_stmt is reworked.

File: src/box/vy_stmt.h [hunk 2]
```diff
@@ -172,7 +172,7 @@ enum {
 struct vy_stmt {
 	struct tuple base;
```
[the remark this answers is not in the document]
Comment (PersDep) [reply src/box/vy_stmt.h:173]: Fixed.

File: src/box/vy_stmt.c [hunk 3]
```diff
@@ -277,13 +281,16 @@ vy_key_new(struct tuple_format *format, const char *key, uint32_t part_count)
 		mp_next(&key_end);
 
 	/* Allocate stmt */
+	bool is_tiny = false;
 	uint32_t key_size = key_end - key;
 	uint32_t bsize = mp_sizeof_array(part_count) + key_size;
-	struct tuple *stmt = vy_stmt_alloc(format, sizeof(struct vy_stmt), bsize);
+	struct tuple *stmt = vy_stmt_alloc(format, sizeof(struct vy_stmt) +
+				!is_tiny * sizeof(uint32_t), bsize, is_tiny);
```
[the remark this answers is not in the document]
Comment (PersDep) [reply src/box/vy_stmt.c:288]: Fixed, no more extra calculations with `is_tiny`.

File: src/box/tuple.h [hunk 4]
```diff
@@ -322,34 +324,147 @@ struct PACKED tuple
...
+			 * Offset to the MessagePack from the beginning of the
+			 * tuple in case is_tiny == false.
+			 */
+			uint16_t data_offset: 15;
+			/**
+			 * Tuple is tiny in case it's bsize fits in 1 byte,
+			 * it's data_offset fits in 6 bits and field map
+			 * offsets fit into 1 byte each.
+			 */
+			bool is_tiny : 1;
```
[the remark this answers is not in the document]
Comment (PersDep) [reply src/box/tuple.h:339]: Fixed using named structs with the same leading fields `is_tiny` which are guaranteed to be the same.

File: src/box/tuple.h [hunk 5]
```diff
@@ -322,34 +324,147 @@ struct PACKED tuple
...
-	 * The tuple (if it's found in index for example) could be invisible
-	 * for current transactions. The flag means that the tuple must
-	 * be clarified by transaction engine.
+	 * Last 31 bits are the length of the MessagePack data
+	 * in the raw part of the tuple.
+	 * First bit contains is_dirty flag.
+	 * Valid in case of is_tiny == false. Otherwise empty.
 	 */
-	bool is_dirty : 1;
+	uint32_t bsize[];
```
[the remark this answers is not in the document]
Comment (PersDep) [reply src/box/tuple.h:366]: Well, yes, to some extent. The variable size struct should not be used as the first member of other structs. It is not anymore.
There might be some other ways to make tuples' size variable (without making `struct tuple` size variable through array of variable size at the end) but they seem to be worse. With the new fixes this one looks good enough.

**Comment by PersDep** [comment]
> 1. The main part of the patch.
>    1.1. I don't like but I can live with `bool is_type` added to lots of functions. But the necessity of tracking the size of tuple manually seems to be unacceptable. See `vy_stmt_set_flags` method, it became a crap. Also I'm afraid of UB.
>    1.2. A least getters and setters of `bsize` should be extracted to separate commit.

The comments on the main part of the patch are answered. Issues are fixed in the new PR: https://github.com/tarantool/tarantool/pull/5608

> 2. The test
>    2.1. There's huge amount of bugs.
>    2.2. The proposed test subsystem (perf.h) is wrong and unusable.
>    2.3. The test itself is chaotic and hard to understand.
>    2.4. It does not test:
>    2.4.1. Performance of accessing tuple members (like bsize) in different workloads - tiny, big and mix.
>    2.4.2. Performance of access to indexed tuple fields. Current test of field access is inappropriate, it tests the speed of  `mp_next`.

Yes, as far as i see, the test should be pretty much reworked. New version will be introduced separately.

**Comment by PersDep** [comment]
https://github.com/tarantool/tarantool/pull/5614

**Comment by PersDep** [comment]
Continued in #5608, #5614 =

**Pull request closed without merging**